File: services/evm/src/block_access_list.rs
//! Block Access List (BAL) construction and hashing
//!
//! Based on EIP-7928 adapted for JAM's stateless verification model.
//! Constructs BAL from verified UBT witnesses for deterministic block_access_list_hash.

use alloc::collections::BTreeMap;
use alloc::vec::Vec;
use primitive_types::{H160, H256};
use crate::ubt::{UBTKeyType, UBTWitnessEntry};

/// Block Access List - all state accesses in a block
#[derive(Debug, Clone)]
pub struct BlockAccessList {
    pub accounts: Vec<AccountChanges>,  // Sorted by address
}

/// Account-level changes
#[derive(Debug, Clone)]
pub struct AccountChanges {
    pub address: H160,
    pub storage_reads: Vec<StorageRead>,      // Read-only slots (sorted by key)
    pub storage_changes: Vec<SlotChanges>,    // Modified slots (sorted by key)
    pub balance_changes: Vec<BalanceChange>,  // Ordered by tx_index
    pub nonce_changes: Vec<NonceChange>,      // Ordered by tx_index
    pub code_changes: Vec<CodeChange>,        // Ordered by tx_index
}

/// Storage slot that was read but not modified
#[derive(Debug, Clone)]
pub struct StorageRead {
    pub key: H256,  // Full 32-byte storage key
}

/// Storage slot that was modified (may have multiple writes)
#[derive(Debug, Clone)]
pub struct SlotChanges {
    pub key: H256,
    pub writes: Vec<SlotWrite>,  // Ordered by tx_index
}

/// Single write to a storage slot
#[derive(Debug, Clone)]
pub struct SlotWrite {
    pub block_access_index: u32,  // Transaction index
    pub pre_value: H256,
    pub post_value: H256,
}

/// Balance modification
#[derive(Debug, Clone)]
pub struct BalanceChange {
    pub block_access_index: u32,
    pub pre_balance: H256,   // Big-endian uint128 in last 16 bytes (BasicData format)
    pub post_balance: H256,
}

/// Nonce modification
#[derive(Debug, Clone)]
pub struct NonceChange {
    pub block_access_index: u32,
    pub pre_nonce: [u8; 8],   // Little-endian uint64 (BasicData format)
    pub post_nonce: [u8; 8],
}

/// Code deployment
#[derive(Debug, Clone)]
pub struct CodeChange {
    pub block_access_index: u32,
    pub code_hash: H256,   // keccak256 of code
    pub code_size: u32,
}

impl BlockAccessList {
    /// Build BAL from verified UBT witness entries.
    pub fn from_ubt_entries(
        read_entries: &[UBTWitnessEntry],
        write_entries: &[UBTWitnessEntry],
    ) -> Self {
        let mut account_map: BTreeMap<H160, AccountChanges> = BTreeMap::new();

        let write_keys = collect_write_keys_ubt(write_entries);

        for read in read_entries {
            if is_modified_ubt(read, &write_keys) {
                continue;
            }

            let account = account_map
                .entry(read.metadata.address)
                .or_insert_with(|| AccountChanges::new(read.metadata.address));

            if read.metadata.key_type == UBTKeyType::Storage {
                account.storage_reads.push(StorageRead {
                    key: read.metadata.storage_key,
                });
            }
        }

        let slot_changes = build_slot_changes_ubt(write_entries);
        let balance_changes = build_balance_changes_ubt(write_entries);
        let nonce_changes = build_nonce_changes_ubt(write_entries);
        let code_changes = build_code_changes_ubt(write_entries);

        for (addr, slots) in slot_changes {
            account_map
                .entry(addr)
                .or_insert_with(|| AccountChanges::new(addr))
                .storage_changes = slots;
        }

        for (addr, changes) in balance_changes {
            account_map
                .entry(addr)
                .or_insert_with(|| AccountChanges::new(addr))
                .balance_changes = changes;
        }

        for (addr, changes) in nonce_changes {
            account_map
                .entry(addr)
                .or_insert_with(|| AccountChanges::new(addr))
                .nonce_changes = changes;
        }

        for (addr, changes) in code_changes {
            account_map
                .entry(addr)
                .or_insert_with(|| AccountChanges::new(addr))
                .code_changes = changes;
        }

        let mut accounts: Vec<AccountChanges> = account_map.into_values().collect();

        for account in &mut accounts {
            account.storage_reads.sort_by_key(|r| r.key);
            account.storage_changes.sort_by_key(|c| c.key);
        }

        BlockAccessList { accounts }
    }

    /// Compute Blake2b hash of RLP-encoded BAL
    pub fn hash(&self) -> [u8; 32] {
        let encoded = self.rlp_encode();
        utils::hash_functions::blake2b_hash(&encoded)
    }

    /// RLP encode the BAL structure
    fn rlp_encode(&self) -> Vec<u8> {
        use rlp::RlpStream;

        // BlockAccessList is a struct with one field: Accounts
        // RLP encoding should wrap the accounts list in an outer list
        let mut stream = RlpStream::new_list(1); // Outer struct wrapper
        stream.begin_list(self.accounts.len());  // Accounts field

        for account in &self.accounts {
            // Encode AccountChanges as RLP list:
            // [address, storage_reads, storage_changes, balance_changes, nonce_changes, code_changes]
            stream.begin_list(6);
            stream.append(&account.address);

            // StorageReads
            stream.begin_list(account.storage_reads.len());
            for read in &account.storage_reads {
                stream.append(&read.key);
            }

            // StorageChanges
            stream.begin_list(account.storage_changes.len());
            for slot in &account.storage_changes {
                stream.begin_list(2);  // [key, writes]
                stream.append(&slot.key);

                stream.begin_list(slot.writes.len());
                for write in &slot.writes {
                    stream.begin_list(3);  // [index, pre, post]
                    stream.append(&write.block_access_index);
                    stream.append(&write.pre_value);
                    stream.append(&write.post_value);
                }
            }

            // BalanceChanges
            stream.begin_list(account.balance_changes.len());
            for change in &account.balance_changes {
                stream.begin_list(3);
                stream.append(&change.block_access_index);
                stream.append(&change.pre_balance);
                stream.append(&change.post_balance);
            }

            // NonceChanges
            stream.begin_list(account.nonce_changes.len());
            for change in &account.nonce_changes {
                stream.begin_list(3);
                stream.append(&change.block_access_index);
                stream.append(&change.pre_nonce.as_ref());
                stream.append(&change.post_nonce.as_ref());
            }

            // CodeChanges
            stream.begin_list(account.code_changes.len());
            for change in &account.code_changes {
                stream.begin_list(3);
                stream.append(&change.block_access_index);
                stream.append(&change.code_hash);
                stream.append(&change.code_size);
            }
        }

        stream.out().to_vec()
    }
}

impl AccountChanges {
    fn new(address: H160) -> Self {
        Self {
            address,
            storage_reads: Vec::new(),
            storage_changes: Vec::new(),
            balance_changes: Vec::new(),
            nonce_changes: Vec::new(),
            code_changes: Vec::new(),
        }
    }
}

fn is_modified_ubt(read: &UBTWitnessEntry, write_keys: &BTreeMap<(H160, H256), ()>) -> bool {
    if read.metadata.key_type == UBTKeyType::Storage {
        write_keys.contains_key(&(read.metadata.address, read.metadata.storage_key))
    } else {
        false
    }
}

fn collect_write_keys_ubt(writes: &[UBTWitnessEntry]) -> BTreeMap<(H160, H256), ()> {
    let mut keys = BTreeMap::new();
    for write in writes {
        if write.metadata.key_type == UBTKeyType::Storage {
            keys.insert((write.metadata.address, write.metadata.storage_key), ());
        }
    }
    keys
}

fn build_slot_changes_ubt(writes: &[UBTWitnessEntry]) -> BTreeMap<H160, Vec<SlotChanges>> {
    let mut slot_map: BTreeMap<H160, BTreeMap<H256, Vec<SlotWrite>>> = BTreeMap::new();

    for write in writes {
        if write.metadata.key_type != UBTKeyType::Storage {
            continue;
        }

        slot_map
            .entry(write.metadata.address)
            .or_insert_with(BTreeMap::new)
            .entry(write.metadata.storage_key)
            .or_insert_with(Vec::new)
            .push(SlotWrite {
                block_access_index: write.metadata.tx_index,
                pre_value: H256::from(write.pre_value),
                post_value: H256::from(write.post_value),
            });
    }

    slot_map
        .into_iter()
        .map(|(addr, slots)| {
            let changes = slots
                .into_iter()
                .map(|(key, writes)| SlotChanges { key, writes })
                .collect();
            (addr, changes)
        })
        .collect()
}

fn build_balance_changes_ubt(writes: &[UBTWitnessEntry]) -> BTreeMap<H160, Vec<BalanceChange>> {
    let mut changes: BTreeMap<H160, Vec<BalanceChange>> = BTreeMap::new();

    for write in writes {
        if write.metadata.key_type != UBTKeyType::BasicData {
            continue;
        }

        let mut pre_balance = [0u8; 32];
        let mut post_balance = [0u8; 32];
        pre_balance[16..32].copy_from_slice(&write.pre_value[16..32]);
        post_balance[16..32].copy_from_slice(&write.post_value[16..32]);

        changes
            .entry(write.metadata.address)
            .or_insert_with(Vec::new)
            .push(BalanceChange {
                block_access_index: write.metadata.tx_index,
                pre_balance: H256::from(pre_balance),
                post_balance: H256::from(post_balance),
            });
    }

    changes
}

fn build_nonce_changes_ubt(writes: &[UBTWitnessEntry]) -> BTreeMap<H160, Vec<NonceChange>> {
    let mut changes: BTreeMap<H160, Vec<NonceChange>> = BTreeMap::new();

    for write in writes {
        if write.metadata.key_type != UBTKeyType::BasicData {
            continue;
        }

        let mut pre_nonce = [0u8; 8];
        let mut post_nonce = [0u8; 8];
        pre_nonce.copy_from_slice(&write.pre_value[8..16]);
        post_nonce.copy_from_slice(&write.post_value[8..16]);

        changes
            .entry(write.metadata.address)
            .or_insert_with(Vec::new)
            .push(NonceChange {
                block_access_index: write.metadata.tx_index,
                pre_nonce,
                post_nonce,
            });
    }

    changes
}

fn build_code_changes_ubt(writes: &[UBTWitnessEntry]) -> BTreeMap<H160, Vec<CodeChange>> {
    let mut code_sizes: BTreeMap<H160, (u32, u32)> = BTreeMap::new();

    for write in writes {
        if write.metadata.key_type == UBTKeyType::BasicData {
            let code_size = ((write.post_value[5] as u32) << 16)
                | ((write.post_value[6] as u32) << 8)
                | (write.post_value[7] as u32);

            if code_size > 0 {
                code_sizes.insert(write.metadata.address, (write.metadata.tx_index, code_size));
            }
        }
    }

    let mut changes: BTreeMap<H160, Vec<CodeChange>> = BTreeMap::new();

    for write in writes {
        if write.metadata.key_type == UBTKeyType::CodeHash {
            let code_hash = H256::from(write.post_value);
            let (tx_index, code_size) = code_sizes
                .get(&write.metadata.address)
                .copied()
                .unwrap_or((write.metadata.tx_index, 0));

            changes
                .entry(write.metadata.address)
                .or_insert_with(Vec::new)
                .push(CodeChange {
                    block_access_index: tx_index,
                    code_hash,
                    code_size,
                });
        }
    }

    changes
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_bal() {
        let bal = BlockAccessList::from_ubt_entries(&[], &[]);
        assert_eq!(bal.accounts.len(), 0);

        // Empty BAL should have deterministic hash
        let hash = bal.hash();
        assert_eq!(hash.len(), 32);
    }

    #[test]
    fn test_bal_hash_determinism() {
        let bal1 = BlockAccessList::from_ubt_entries(&[], &[]);
        let bal2 = BlockAccessList::from_ubt_entries(&[], &[]);

        assert_eq!(bal1.hash(), bal2.hash());
    }
}

File: storage/block_access_list.go
package storage

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"sort"

	"github.com/colorfulnotion/jam/common"
	"github.com/ethereum/go-ethereum/rlp"
)

// Block Access List (BAL) implementation for JAM EVM
// Based on EIP-7928 adapted for JAM's stateless verification model
//
// Design: Guarantor reconstructs BAL from augmented witness (161-byte entries with txIndex)
// without re-execution, ensuring deterministic block_access_list_hash matching.

// ===== Data Structures =====

// BlockAccessList represents all state accesses in a block
type BlockAccessList struct {
	Accounts []AccountChanges // Sorted by address bytes
}

// AccountChanges tracks all accesses to a single account
type AccountChanges struct {
	Address        common.Address
	StorageReads   []StorageRead   // Read-only storage slots (sorted by key)
	StorageChanges []SlotChanges   // Modified storage slots (sorted by key)
	BalanceChanges []BalanceChange // Ordered by blockAccessIndex
	NonceChanges   []NonceChange   // Ordered by blockAccessIndex
	CodeChanges    []CodeChange    // Ordered by blockAccessIndex
}

// StorageRead represents a read-only storage access
type StorageRead struct {
	Key common.Hash // Full storage key
}

// SlotChanges tracks all writes to a single storage slot
type SlotChanges struct {
	Key    common.Hash
	Writes []SlotWrite // Ordered by blockAccessIndex
}

// SlotWrite represents a single write to a storage slot
type SlotWrite struct {
	BlockAccessIndex uint32
	PreValue         common.Hash // Fixed-size for deterministic RLP
	PostValue        common.Hash
}

// BalanceChange represents a balance modification
type BalanceChange struct {
	BlockAccessIndex uint32
	PreBalance       common.Hash // Big-endian uint128 in last 16 bytes
	PostBalance      common.Hash
}

// NonceChange represents a nonce modification
type NonceChange struct {
	BlockAccessIndex uint32
	PreNonce         [8]byte // Little-endian uint64
	PostNonce        [8]byte
}

// CodeChange represents code deployment
type CodeChange struct {
	BlockAccessIndex uint32
	CodeHash         common.Hash // keccak256 of code
	CodeSize         uint32
}

// ===== Hash Computation =====

// Hash computes deterministic block_access_list_hash using RLP encoding + Blake2b
// Note: Uses Blake2b (JAM standard) not Keccak256 (EIP-7928 uses Keccak for Ethereum)
func (bal *BlockAccessList) Hash() common.Hash {
	encoded, err := rlp.EncodeToBytes(bal)
	if err != nil {
		panic(fmt.Sprintf("BAL RLP encoding failed: %v", err))
	}
	return common.Blake2Hash(encoded)
}

// ===== Witness Parsing =====

// ReadEntry represents a single entry from the pre-state witness section
type ReadEntry struct {
	UBTKey     common.Hash
	KeyType    uint8
	Address    common.Address
	Extra      uint64
	StorageKey common.Hash
	PreValue   common.Hash
	PostValue  common.Hash // Hint only (same as PreValue for reads)
	TxIndex    uint32      // Transaction index (blockAccessIndex)
}

// WriteEntry represents a single entry from the post-state witness section
type WriteEntry struct {
	UBTKey     common.Hash
	KeyType    uint8
	Address    common.Address
	Extra      uint64
	StorageKey common.Hash
	PreValue   common.Hash
	PostValue  common.Hash // Authoritative value from contract witness
	TxIndex    uint32      // Transaction index (blockAccessIndex)
}

// parseAugmentedPreStateReads parses the pre-state section of augmented witness
// Format: [32B pre_root][4B count][161B entries...][4B proof_len][proof...]
func parseAugmentedPreStateReads(preStateWitness []byte) ([]ReadEntry, error) {
	if len(preStateWitness) < 36 {
		return nil, fmt.Errorf("pre-state witness too short: %d bytes", len(preStateWitness))
	}

	offset := 32 // Skip pre_root
	count := binary.BigEndian.Uint32(preStateWitness[offset : offset+4])
	offset += 4

	requiredSize := offset + int(count)*161 + 4 // entries + proof_len
	if len(preStateWitness) < requiredSize {
		return nil, fmt.Errorf("pre-state witness truncated: need %d, got %d", requiredSize, len(preStateWitness))
	}

	reads := make([]ReadEntry, count)
	for i := uint32(0); i < count; i++ {
		entry := &reads[i]

		// 32B UBTKey
		copy(entry.UBTKey[:], preStateWitness[offset:offset+32])
		offset += 32

		// 1B KeyType
		entry.KeyType = preStateWitness[offset]
		offset += 1

		// 20B Address
		copy(entry.Address[:], preStateWitness[offset:offset+20])
		offset += 20

		// 8B Extra
		entry.Extra = binary.BigEndian.Uint64(preStateWitness[offset : offset+8])
		offset += 8

		// 32B StorageKey
		copy(entry.StorageKey[:], preStateWitness[offset:offset+32])
		offset += 32

		// 32B PreValue
		copy(entry.PreValue[:], preStateWitness[offset:offset+32])
		offset += 32

		// 32B PostValue (same as PreValue for reads)
		copy(entry.PostValue[:], preStateWitness[offset:offset+32])
		offset += 32

		// 4B TxIndex
		entry.TxIndex = binary.BigEndian.Uint32(preStateWitness[offset : offset+4])
		offset += 4
	}

	return reads, nil
}

// parseAugmentedPostStateWrites parses the post-state section of augmented witness
// Format: [32B post_root][4B count][161B entries...][4B proof_len][proof...]
func parseAugmentedPostStateWrites(postStateWitness []byte) ([]WriteEntry, error) {
	if len(postStateWitness) < 36 {
		return nil, fmt.Errorf("post-state witness too short: %d bytes", len(postStateWitness))
	}

	offset := 32 // Skip post_root
	count := binary.BigEndian.Uint32(postStateWitness[offset : offset+4])
	offset += 4

	requiredSize := offset + int(count)*161 + 4 // entries + proof_len
	if len(postStateWitness) < requiredSize {
		return nil, fmt.Errorf("post-state witness truncated: need %d, got %d", requiredSize, len(postStateWitness))
	}

	writes := make([]WriteEntry, count)
	for i := uint32(0); i < count; i++ {
		entry := &writes[i]

		// 32B UBTKey
		copy(entry.UBTKey[:], postStateWitness[offset:offset+32])
		offset += 32

		// 1B KeyType
		entry.KeyType = postStateWitness[offset]
		offset += 1

		// 20B Address
		copy(entry.Address[:], postStateWitness[offset:offset+20])
		offset += 20

		// 8B Extra
		entry.Extra = binary.BigEndian.Uint64(postStateWitness[offset : offset+8])
		offset += 8

		// 32B StorageKey
		copy(entry.StorageKey[:], postStateWitness[offset:offset+32])
		offset += 32

		// 32B PreValue
		copy(entry.PreValue[:], postStateWitness[offset:offset+32])
		offset += 32

		// 32B PostValue
		copy(entry.PostValue[:], postStateWitness[offset:offset+32])
		offset += 32

		// 4B TxIndex
		entry.TxIndex = binary.BigEndian.Uint32(postStateWitness[offset : offset+4])
		offset += 4
	}

	return writes, nil
}

// ===== BAL Construction =====

const (
	KeyTypeBasicData = 0
	KeyTypeCodeHash  = 1
	KeyTypeCodeChunk = 2
	KeyTypeStorage   = 3
)

// buildSlotChanges groups storage writes by address and key, ordered by blockAccessIndex
func buildSlotChanges(writes []WriteEntry) map[common.Address]map[common.Hash][]SlotWrite {
	slotMap := make(map[common.Address]map[common.Hash][]SlotWrite)

	for _, write := range writes {
		if write.KeyType != KeyTypeStorage {
			continue
		}

		if slotMap[write.Address] == nil {
			slotMap[write.Address] = make(map[common.Hash][]SlotWrite)
		}

		slotMap[write.Address][write.StorageKey] = append(
			slotMap[write.Address][write.StorageKey],
			SlotWrite{
				BlockAccessIndex: write.TxIndex,
				PreValue:         write.PreValue,
				PostValue:        write.PostValue,
			},
		)
	}

	return slotMap
}

// buildBalanceChanges extracts balance changes from BasicData writes
func buildBalanceChanges(writes []WriteEntry) map[common.Address][]BalanceChange {
	changes := make(map[common.Address][]BalanceChange)

	for _, write := range writes {
		if write.KeyType != KeyTypeBasicData {
			continue
		}

		// Extract balance from BasicData (bytes 16-31)
		var preBalance, postBalance common.Hash
		copy(preBalance[16:], write.PreValue[16:])
		copy(postBalance[16:], write.PostValue[16:])

		changes[write.Address] = append(changes[write.Address], BalanceChange{
			BlockAccessIndex: write.TxIndex,
			PreBalance:       preBalance,
			PostBalance:      postBalance,
		})
	}

	return changes
}

// buildNonceChanges extracts nonce changes from BasicData writes
func buildNonceChanges(writes []WriteEntry) map[common.Address][]NonceChange {
	changes := make(map[common.Address][]NonceChange)

	for _, write := range writes {
		if write.KeyType != KeyTypeBasicData {
			continue
		}

		// Extract nonce from BasicData (bytes 8-15, little-endian)
		var preNonce, postNonce [8]byte
		copy(preNonce[:], write.PreValue[8:16])
		copy(postNonce[:], write.PostValue[8:16])

		changes[write.Address] = append(changes[write.Address], NonceChange{
			BlockAccessIndex: write.TxIndex,
			PreNonce:         preNonce,
			PostNonce:        postNonce,
		})
	}

	return changes
}

// buildCodeChanges extracts code deployments from CodeHash writes
// Cross-references BasicData writes to extract code size from bytes 5-7 (3-byte big-endian uint24)
func buildCodeChanges(writes []WriteEntry) map[common.Address][]CodeChange {
	changes := make(map[common.Address][]CodeChange)

	// First pass: Extract code sizes from BasicData writes (offset 5-7)
	codeSizeMap := make(map[common.Address]uint32)
	for _, write := range writes {
		if write.KeyType != KeyTypeBasicData {
			continue
		}

		// BasicData PostValue contains code_size at bytes 5-7 (3-byte big-endian)
		// Per EIP-6800: offset 0=version, 5-7=code_size, 8-15=nonce, 16-31=balance
		postValue := write.PostValue[:]
		codeSize := (uint32(postValue[5]) << 16) | (uint32(postValue[6]) << 8) | uint32(postValue[7])
		codeSizeMap[write.Address] = codeSize
	}

	// Second pass: Build CodeChange entries with code size from BasicData
	for _, write := range writes {
		if write.KeyType != KeyTypeCodeHash {
			continue
		}

		// Get code size from BasicData cross-reference (default 0 if not found)
		codeSize := codeSizeMap[write.Address]

		// PostValue contains the code hash
		changes[write.Address] = append(changes[write.Address], CodeChange{
			BlockAccessIndex: write.TxIndex,
			CodeHash:         write.PostValue,
			CodeSize:         codeSize,
		})
	}

	return changes
}

// classifyStorageAccesses separates read-only storage accesses from modified ones
func classifyStorageAccesses(
	reads []ReadEntry,
	writes []WriteEntry,
) map[common.Address][]StorageRead {
	// Build write set for quick lookup
	writeSet := make(map[common.Address]map[common.Hash]struct{})
	for _, write := range writes {
		if write.KeyType != KeyTypeStorage {
			continue
		}
		if writeSet[write.Address] == nil {
			writeSet[write.Address] = make(map[common.Hash]struct{})
		}
		writeSet[write.Address][write.StorageKey] = struct{}{}
	}

	// Classify reads as read-only if not written
	readOnly := make(map[common.Address][]StorageRead)
	for _, read := range reads {
		if read.KeyType != KeyTypeStorage {
			continue
		}

		// Check if this key was written
		if writeSet[read.Address] != nil {
			if _, written := writeSet[read.Address][read.StorageKey]; written {
				continue // Was written, not read-only
			}
		}

		readOnly[read.Address] = append(readOnly[read.Address], StorageRead{
			Key: read.StorageKey,
		})
	}

	return readOnly
}

// assembleAccountChanges merges all change maps into sorted AccountChanges list
func assembleAccountChanges(
	storageReadsMap map[common.Address][]StorageRead,
	slotChangesMap map[common.Address]map[common.Hash][]SlotWrite,
	balanceChangesMap map[common.Address][]BalanceChange,
	nonceChangesMap map[common.Address][]NonceChange,
	codeChangesMap map[common.Address][]CodeChange,
) []AccountChanges {
	// Collect all touched addresses
	addressSet := make(map[common.Address]struct{})
	for addr := range storageReadsMap {
		addressSet[addr] = struct{}{}
	}
	for addr := range slotChangesMap {
		addressSet[addr] = struct{}{}
	}
	for addr := range balanceChangesMap {
		addressSet[addr] = struct{}{}
	}
	for addr := range nonceChangesMap {
		addressSet[addr] = struct{}{}
	}
	for addr := range codeChangesMap {
		addressSet[addr] = struct{}{}
	}

	// Sort addresses lexicographically
	addresses := make([]common.Address, 0, len(addressSet))
	for addr := range addressSet {
		addresses = append(addresses, addr)
	}
	sort.Slice(addresses, func(i, j int) bool {
		return bytes.Compare(addresses[i][:], addresses[j][:]) < 0
	})

	// Build AccountChanges per address
	accounts := make([]AccountChanges, 0, len(addresses))
	for _, addr := range addresses {
		// Sort storage reads by key
		storageReads := storageReadsMap[addr]
		sort.Slice(storageReads, func(i, j int) bool {
			return bytes.Compare(storageReads[i].Key[:], storageReads[j].Key[:]) < 0
		})

		// Build sorted SlotChanges
		slotMap := slotChangesMap[addr]
		storageKeys := make([]common.Hash, 0, len(slotMap))
		for key := range slotMap {
			storageKeys = append(storageKeys, key)
		}
		sort.Slice(storageKeys, func(i, j int) bool {
			return bytes.Compare(storageKeys[i][:], storageKeys[j][:]) < 0
		})

		slotChanges := make([]SlotChanges, 0, len(storageKeys))
		for _, key := range storageKeys {
			writes := slotMap[key]
			// Sort writes by blockAccessIndex
			sort.Slice(writes, func(i, j int) bool {
				return writes[i].BlockAccessIndex < writes[j].BlockAccessIndex
			})
			slotChanges = append(slotChanges, SlotChanges{
				Key:    key,
				Writes: writes,
			})
		}

		// Sort balance/nonce/code changes by blockAccessIndex
		balanceChanges := balanceChangesMap[addr]
		sort.Slice(balanceChanges, func(i, j int) bool {
			return balanceChanges[i].BlockAccessIndex < balanceChanges[j].BlockAccessIndex
		})

		nonceChanges := nonceChangesMap[addr]
		sort.Slice(nonceChanges, func(i, j int) bool {
			return nonceChanges[i].BlockAccessIndex < nonceChanges[j].BlockAccessIndex
		})

		codeChanges := codeChangesMap[addr]
		sort.Slice(codeChanges, func(i, j int) bool {
			return codeChanges[i].BlockAccessIndex < codeChanges[j].BlockAccessIndex
		})

		accounts = append(accounts, AccountChanges{
			Address:        addr,
			StorageReads:   storageReads,
			StorageChanges: slotChanges,
			BalanceChanges: balanceChanges,
			NonceChanges:   nonceChanges,
			CodeChanges:    codeChanges,
		})
	}

	return accounts
}

// BuildBlockAccessList constructs BAL from augmented UBT witness
//
// Process:
// 1. Parse augmented pre-state witness for reads (161-byte entries with txIndex)
// 2. Parse augmented post-state witness for writes (161-byte entries with txIndex)
// 3. Build changes using txIndex as blockAccessIndex
// 4. Classify into read-only vs modified
// 5. Sort and deduplicate
// 6. Return BlockAccessList with proper ordering
//
// Note: Uses witness metadata only - guarantor can call this without execution state.
func BuildBlockAccessList(
	preStateWitness []byte,
	postStateWitness []byte,
) (*BlockAccessList, error) {
	// Parse augmented witness entries
	reads, err := parseAugmentedPreStateReads(preStateWitness)
	if err != nil {
		return nil, fmt.Errorf("failed to parse pre-state witness: %w", err)
	}

	writes, err := parseAugmentedPostStateWrites(postStateWitness)
	if err != nil {
		return nil, fmt.Errorf("failed to parse post-state witness: %w", err)
	}

	// Build changes from witness entries
	slotChangesMap := buildSlotChanges(writes)
	balanceChangesMap := buildBalanceChanges(writes)
	nonceChangesMap := buildNonceChanges(writes)
	codeChangesMap := buildCodeChanges(writes)

	// Classify storage reads
	storageReadsMap := classifyStorageAccesses(reads, writes)

	// Assemble AccountChanges with deterministic ordering
	accounts := assembleAccountChanges(
		storageReadsMap,
		slotChangesMap,
		balanceChangesMap,
		nonceChangesMap,
		codeChangesMap,
	)

	return &BlockAccessList{Accounts: accounts}, nil
}

// splitWitnessSections parses the dual-proof witness into pre/post sections
func SplitWitnessSections(witness []byte) (preState, postState []byte, err error) {
	// Parse witness structure:
	// [32B pre_root][4B read_count][read_entries...][4B pre_proof_len][pre_proof...]
	// [32B post_root][4B write_count][write_entries...][4B post_proof_len][post_proof...]

	if len(witness) < 36 {
		return nil, nil, fmt.Errorf("witness too short: %d bytes", len(witness))
	}

	offset := 0

	// Read count
	readCount := binary.BigEndian.Uint32(witness[32:36])
	offset = 36 + int(readCount)*161 // 161 bytes per entry

	// Pre-proof length
	if offset+4 > len(witness) {
		return nil, nil, fmt.Errorf("invalid witness: missing pre_proof_len at offset %d", offset)
	}
	preProofLen := binary.BigEndian.Uint32(witness[offset : offset+4])
	offset += 4 + int(preProofLen)

	if offset+4 > len(witness) {
		return nil, nil, fmt.Errorf("invalid witness: missing pre_extension_count at offset %d", offset)
	}
	extCount := binary.BigEndian.Uint32(witness[offset : offset+4])
	offset += 4
	extSize := int(extCount) * (32 + 31 + 32 + 2 + 248*32)
	offset += extSize

	if offset > len(witness) {
		return nil, nil, fmt.Errorf("invalid witness: pre-section overflow at offset %d", offset)
	}

	// Pre-state section ends here
	preState = witness[:offset]

	// Post-state section starts here
	postState = witness[offset:]

	return preState, postState, nil
}
